### PR TITLE
Add LCE model optimisation guide.

### PR DESCRIPTION
**Pull request by AdamHillier**

### CODE REVIEW

File: docs/compute-engine/model_optimization_guide.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    x = lq.layers.QuantDense(num_dense_outputs, ...)(x)
+
+    # ...with this
+    x = tf.keras.layers.Reshape((1, 1, -1))(x)
+    x = lq.layers.QuantConv2D(num_dense_outputs, kernel_size=(1, 1), ...)(x)
+    x = tf.keras.layers.Reshape((-1,))(x)
+    ```
+
+*   **Ensure that `filter_height` × `filter_width` × `input_channels` is a
+    multiple of 128**
```
Comment (Tombana) [inline docs/compute-engine/model_optimization_guide.md:101]: For the common filter sizes `1x1`, `3x3` and `5x5` (or any filter of which the number of pixels isn't divisible by two), to only way to achieve this constraint is to use multiples of 128 input channels, which is quite a strong constraint I think. Therefore I'm a bit unsure about how this part should be phrased. Maybe we can say that this optimization is less important than the one about multiples of 32 channels. (The multiple of 32 affects bitpacking, im2col, ruy packing and bgemm, whereas this particular constraint only affects ruy packing and bgemm)

**Review by Tombana — APPROVED** [review]
Looks great to me!

File: docs/compute-engine/model_optimization_guide.md (hunk 1 above)
Comment (AdamHillier) [reply docs/compute-engine/model_optimization_guide.md:101]: Yeah agreed, I'll try and think about how to reword this

Comment (Tombana) [reply docs/compute-engine/model_optimization_guide.md:101]: @koenhelwegen Have you observed this effect in your benchmarks? i.e. a `1x1` convolution on 64 input channels should be only slightly faster than on 128 input channels instead of 2X faster (bitpacking and im2col should be 2X faster, whereas the main bgemm part should be equally fast).

File: docs/compute-engine/model_optimization_guide.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    input with 32 channels will run equally fast.
+
+*   **Use multiples of 8 output channels**
+
+    The LCE optimised binary convolution op computes the result of 8 output
+    channels in parallel, so if the number of output channels is not a multiple
+    of 8 there will be wasted computation.
+
+*   **Use `QuantConv2D` → `ReLU` → `BatchNormalization` to take advantage of op
+    fusing**
```
Comment (lgeiger) [inline docs/compute-engine/model_optimization_guide.md:27]: How do you feel about changing the title to something like this:
```suggestion
*   **Use `QuantConv2D` → `ReLU` → `BatchNormalization` instead of `QuantConv2D` → `BatchNormalization` → `ReLU` to take advantage of op fusing**
```

While the text explains it great, to me the title suggests that I should always use a relu between bconv and batch norm but it would still be fine to not use a relu at all.

**Review by lgeiger — APPROVED** [review]
This is excellent, thanks for writing the guide!

File: docs/compute-engine/model_optimization_guide.md (hunk 1 above)
Comment (koenhelwegen) [reply docs/compute-engine/model_optimization_guide.md:101]: The script currently only allows same input and output channels, but it seems to scale pretty quadratically in channels, so presumably linear in input channels (width/height = 8, kernel_size=1, bitpacked in/out = True):

| in/out channels | block latency (ms) | im2col (ms) | bgemv (ms) |
| --- | --- | --- | --- |
| 64 | 2.28 | 0.0885 | 2.195125 |
| 128 | 8.07 | 0.107 | 7.932125 |

File: docs/compute-engine/model_optimization_guide.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    # Replace this...
+    x = lq.layers.QuantDense(num_dense_outputs, ...)(x)
+
+    # ...with this
+    x = tf.keras.layers.Reshape((1, 1, -1))(x)
+    x = lq.layers.QuantConv2D(num_dense_outputs, kernel_size=(1, 1), ...)(x)
+    x = tf.keras.layers.Reshape((-1,))(x)
+    ```
+
+*   **Ensure that `filter_height` × `filter_width` × `input_channels` is a
```
Comment (jneeven) [inline docs/compute-engine/model_optimization_guide.md:100]:
```suggestion
*   **Ensure that `kernel_height` × `kernel_width` × `input_channels` is a
```

I think this makes it easier to understand which part of the code this is referring to, but I don't feel too strongly about this

File: docs/compute-engine/model_optimization_guide.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    multiple of 128**
+
+    Inside the optimised binary convolution op, an `im2col` procedure is applied
+    to the input tensor, which reduces the convolution operation to a (binary)
+    matrix multiplication. The `im2col`-transformed input has a depth of size
+    `filter_height` × `filter_width` × `input_channels`, which is padded to a
```
Comment (jneeven) [inline docs/compute-engine/model_optimization_guide.md:106]: The same applies here, btw I think these detailed arguments explaining *why* these best practices are necessary are very nice :+1:

**Review by jneeven — APPROVED** [review]
Very nice, thanks :+1:

File: docs/compute-engine/model_optimization_guide.md (hunk 1 above)
Comment (Tombana) [reply docs/compute-engine/model_optimization_guide.md:101]: Without this constraint, we'd expect the 64 case to be 4X faster than the 128 case, so it would be about 2.0 ms, but due to this padding effect we expect it to be anywhere between 2.0 and 8.0 ms. It appears that its about 2.3 ms so apparently this effect is very small. I would have expected a bigger impact to be honest, but I guess this is better. Maybe we can say that we measured the impact of this and that it definitely is measurable but it's not very strong?

Comment (koenhelwegen) [reply docs/compute-engine/model_optimization_guide.md:101]: Yes this seems to be true on pixel1. For a 1x1 kernel we get the following:

| width/height | in/out channels | latency (ms) |
| --- | --- | --- |
| 8 | 32 | 0.008 |
| 8 | 128 | 0.020 |
| 32 | 32 | 0.103 |
| 32 | 128 | 0.298 |

A naive estimate would indicate the 128 channel blocks should be 16x slower, but it's less than a factor 4 in both cases.

Comment (Tombana) [reply docs/compute-engine/model_optimization_guide.md]: The effect seems to be pretty strong then. However usually binary convs are `3x3` in which case the `9 * 32` gets rounded up to `12 * 32` (33% extra for 32 input channels) and `9 * 64` gets rounded to `10 * 64` (11% extra for 64 input channels).

So for `1x1` kernels the effect is strong (2X for 64 channels, 4X for 32 channels), but for `3x3` kernels the effect would be 11% instead of 2X and 33% instead of 4X.
